overlord/devicestate/devicestate_test.go

```diff
@@ -2817,3 +2838,446 @@ func (s *deviceMgrSuite) TestDeviceCtxProvided(c *C) {
+		called = true
+		return gadget.ErrNoUpdate
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Check(chg.Err(), IsNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(called, Equals, true)
+	c.Check(s.restartRequests, HasLen, 0)
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreDoUndo(c *C) {
+	var updateCalled bool
+	var updateRollbackDir string
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		updateRollbackDir = path
+		updateCalled = true
+		return nil
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	s.state.Lock()
+
+	terr := s.state.NewTask("error-trigger", "provoking total undo")
+	terr.WaitFor(t)
+	chg.AddTask(terr)
+
+	s.state.Unlock()
+
+	for i := 0; i < 6; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Assert(chg.Err(), NotNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(terr.Status(), Equals, state.ErrorStatus)
+	c.Check(updateCalled, Equals, true)
+	c.Check(updateRollbackDir, Equals, filepath.Join(dirs.SnapRollbackDir, "foo-gadget_34"))
+	// only one request for task do
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystem})
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreRollbackDirExistsRemovedByUpdate(c *C) {
+	var updateCalled bool
+	var passedRollbackDir string
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		updateCalled = true
+		passedRollbackDir = path
+		c.Assert(osutil.IsDirectory(path), Equals, true)
+
+		err := os.RemoveAll(path)
+		c.Assert(err, IsNil)
+
+		return nil
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	rollbackDir := filepath.Join(dirs.SnapRollbackDir, "foo-gadget_34")
+	err := os.MkdirAll(rollbackDir, 0755)
+	c.Assert(err, IsNil)
+
+	for i := 0; i < 6; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Assert(chg.IsReady(), Equals, true)
+	c.Check(chg.Err(), IsNil)
+	c.Check(t.Status(), Equals, state.DoneStatus)
+	c.Check(updateCalled, Equals, true)
+	c.Check(rollbackDir, Equals, passedRollbackDir)
+	// removed in the update, no error from from task handler
+	c.Check(osutil.IsDirectory(rollbackDir), Equals, false)
+	c.Check(s.restartRequests, DeepEquals, []state.RestartType{state.RestartSystem})
+}
+
+func (s *deviceMgrSuite) TestUpdateGadgetOnCoreRollbackDirCreateFailed(c *C) {
+	restore := devicestate.MockGadgetUpdate(func(current, update *gadget.Info, path string) error {
+		c.Fatal("should not have been called")
+		return nil
+	})
+	defer restore()
+
+	chg, t := setupGadgetUpdate(c, s.state)
+
+	rollbackDir := filepath.Join(dirs.SnapRollbackDir, "foo-gadget_34")
+	err := os.MkdirAll(dirs.SnapRollbackDir, 0000)
```

Comment:
This test will fail if we run the tests as root (because who cares about permissions when you can be r00t! ;)
Its still a useful test so maybe at the beginning of the test something like:
```
if os.Geteuid() == 0 {
                c.Skip("this test cannot run as root (permissions are not honored thenl)")
        }
``` 
?